File: python-version/file_transfer/rs_codec.py
"""Reed-Solomon support for the File Transfer JPEG v2 profile.

The profile uses systematic RS(255, 179) over GF(256), primitive polynomial
0x11d, primitive element 2, and generator roots alpha**0 .. alpha**75.
Codeword bytes are polynomial coefficients in descending degree order.
"""

from __future__ import annotations


PRIMITIVE_POLYNOMIAL = 0x11D
GENERATOR_BASE = 0
DATA_BYTES = 179
PARITY_BYTES = 76
CODEWORD_BYTES = DATA_BYTES + PARITY_BYTES
MAX_CORRECTABLE_ERRORS = PARITY_BYTES // 2


class ReedSolomonError(ValueError):
    """Raised when a codeword cannot be corrected."""


_EXP = [0] * 512
_LOG = [0] * 256
_value = 1
for _index in range(255):
    _EXP[_index] = _value
    _LOG[_value] = _index
    _value <<= 1
    if _value & 0x100:
        _value ^= PRIMITIVE_POLYNOMIAL
for _index in range(255, 512):
    _EXP[_index] = _EXP[_index - 255]


def _mul(left: int, right: int) -> int:
    if left == 0 or right == 0:
        return 0
    return _EXP[_LOG[left] + _LOG[right]]


def _div(left: int, right: int) -> int:
    if right == 0:
        raise ZeroDivisionError("GF(256) division by zero")
    if left == 0:
        return 0
    return _EXP[(_LOG[left] - _LOG[right]) % 255]


def _pow_alpha(power: int) -> int:
    return _EXP[power % 255]


def _poly_mul(left: list[int], right: list[int]) -> list[int]:
    result = [0] * (len(left) + len(right) - 1)
    for i, a in enumerate(left):
        if a:
            for j, b in enumerate(right):
                if b:
                    result[i + j] ^= _mul(a, b)
    return result


def _poly_eval_desc(poly: bytes | bytearray | list[int], value: int) -> int:
    result = 0
    for coefficient in poly:
        result = _mul(result, value) ^ coefficient
    return result


def _poly_eval_asc(poly: list[int], value: int) -> int:
    result = 0
    for coefficient in reversed(poly):
        result = _mul(result, value) ^ coefficient
    return result


def _generator() -> list[int]:
    result = [1]
    for index in range(PARITY_BYTES):
        result = _poly_mul(result, [1, _pow_alpha(GENERATOR_BASE + index)])
    return result


_GENERATOR = _generator()


def encode_block(data: bytes) -> bytes:
    """Encode at most 179 bytes, padding the final data block with zeroes."""

    if len(data) > DATA_BYTES:
        raise ValueError(f"RS block exceeds {DATA_BYTES} data bytes")
    message = bytes(data).ljust(DATA_BYTES, b"\0")
    work = bytearray(message + bytes(PARITY_BYTES))
    for index in range(DATA_BYTES):
        coefficient = work[index]
        if coefficient:
            for generator_index in range(1, len(_GENERATOR)):
                work[index + generator_index] ^= _mul(
                    _GENERATOR[generator_index], coefficient
                )
    return message + bytes(work[-PARITY_BYTES:])


def _syndromes(codeword: bytes | bytearray) -> list[int]:
    return [
        _poly_eval_desc(codeword, _pow_alpha(GENERATOR_BASE + index))
        for index in range(PARITY_BYTES)
    ]


def _error_locator(syndromes: list[int]) -> list[int]:
    # Berlekamp-Massey. Coefficients are returned in ascending degree order.
    locator = [1] + [0] * PARITY_BYTES
    previous = [1] + [0] * PARITY_BYTES
    degree = 0
    shift = 1
    previous_discrepancy = 1

    for index in range(PARITY_BYTES):
        discrepancy = syndromes[index]
        for offset in range(1, degree + 1):
            discrepancy ^= _mul(locator[offset], syndromes[index - offset])
        if discrepancy == 0:
            shift += 1
            continue

        saved = locator.copy()
        scale = _div(discrepancy, previous_discrepancy)
        for coefficient_index in range(PARITY_BYTES + 1 - shift):
            if previous[coefficient_index]:
                locator[coefficient_index + shift] ^= _mul(
                    scale, previous[coefficient_index]
                )
        if 2 * degree <= index:
            degree = index + 1 - degree
            previous = saved
            previous_discrepancy = discrepancy
            shift = 1
        else:
            shift += 1

    if degree > MAX_CORRECTABLE_ERRORS:
        raise ReedSolomonError("too many RS symbol errors")
    return locator[: degree + 1]


def _find_error_positions(locator: list[int], length: int) -> list[int]:
    positions: list[int] = []
    for position in range(length):
        degree = length - 1 - position
        inverse_location = _pow_alpha(-degree)
        if _poly_eval_asc(locator, inverse_location) == 0:
            positions.append(position)
    if len(positions) != len(locator) - 1:
        raise ReedSolomonError("RS error locator roots are incomplete")
    return positions


def _solve_magnitudes(
    syndromes: list[int], positions: list[int], length: int
) -> list[int]:
    count = len(positions)
    if count == 0:
        return []
    # Solve the first `count` syndrome equations over GF(256). For root base
    # zero, S_i = sum(error_j * X_j**i).
    matrix: list[list[int]] = []
    locations = [_pow_alpha(length - 1 - position) for position in positions]
    for row_index in range(count):
        row = []
        for location in locations:
            row.append(1 if row_index == 0 else _EXP[(_LOG[location] * row_index) % 255])
        row.append(syndromes[row_index])
        matrix.append(row)

    for column in range(count):
        pivot = next(
            (row for row in range(column, count) if matrix[row][column]), None
        )
        if pivot is None:
            raise ReedSolomonError("singular RS magnitude system")
        matrix[column], matrix[pivot] = matrix[pivot], matrix[column]
        inverse = _div(1, matrix[column][column])
        for item in range(column, count + 1):
            matrix[column][item] = _mul(matrix[column][item], inverse)
        for row in range(count):
            if row == column or matrix[row][column] == 0:
                continue
            scale = matrix[row][column]
            for item in range(column, count + 1):
                matrix[row][item] ^= _mul(scale, matrix[column][item])
    return [matrix[index][count] for index in range(count)]


def decode_block(codeword: bytes) -> bytes:
    """Correct one 255-byte codeword and return its 179 data bytes."""

    if len(codeword) != CODEWORD_BYTES:
        raise ValueError(f"RS codeword must be {CODEWORD_BYTES} bytes")
    corrected = bytearray(codeword)
    syndromes = _syndromes(corrected)
    if not any(syndromes):
        return bytes(corrected[:DATA_BYTES])
    locator = _error_locator(syndromes)
    positions = _find_error_positions(locator, len(corrected))
    magnitudes = _solve_magnitudes(syndromes, positions, len(corrected))
    for position, magnitude in zip(positions, magnitudes):
        corrected[position] ^= magnitude
    if any(_syndromes(corrected)):
        raise ReedSolomonError("RS correction did not produce a valid codeword")
    return bytes(corrected[:DATA_BYTES])


def encode_interleaved(data: bytes) -> tuple[bytes, int]:
    """RS-encode data and interleave complete codewords by byte column."""

    if not data:
        return b"", 0
    blocks = [
        encode_block(data[offset : offset + DATA_BYTES])
        for offset in range(0, len(data), DATA_BYTES)
    ]
    output = bytearray(len(blocks) * CODEWORD_BYTES)
    cursor = 0
    for column in range(CODEWORD_BYTES):
        for block in blocks:
            output[cursor] = block[column]
            cursor += 1
    return bytes(output), len(blocks)


def decode_interleaved(data: bytes, body_length: int, codewords: int) -> bytes:
    """Undo column interleaving, correct each codeword, and trim padding."""

    if body_length < 0 or codewords < 0:
        raise ValueError("negative RS length")
    expected_codewords = (body_length + DATA_BYTES - 1) // DATA_BYTES
    if codewords != expected_codewords:
        raise ValueError("RS codeword count does not match body length")
    if len(data) != codewords * CODEWORD_BYTES:
        raise ValueError("interleaved RS length mismatch")
    if codewords == 0:
        return b""
    blocks = [bytearray(CODEWORD_BYTES) for _ in range(codewords)]
    cursor = 0
    for column in range(CODEWORD_BYTES):
        for block in blocks:
            block[column] = data[cursor]
            cursor += 1
    decoded = b"".join(decode_block(bytes(block)) for block in blocks)
    return decoded[:body_length]


__all__ = [
    "PRIMITIVE_POLYNOMIAL",
    "GENERATOR_BASE",
    "DATA_BYTES",
    "PARITY_BYTES",
    "CODEWORD_BYTES",
    "MAX_CORRECTABLE_ERRORS",
    "ReedSolomonError",
    "encode_block",
    "decode_block",
    "encode_interleaved",
    "decode_interleaved",
]

File: node-version/src/rs-codec.js
'use strict';

// Shortened use of the conventional RS(255, 179) code over GF(2^8).
// Symbols are ordered from the highest to the lowest polynomial power.
const PRIMITIVE = 0x11d;
const FIELD_SIZE = 255;
const DATA_SYMBOLS = 179;
const PARITY_SYMBOLS = FIELD_SIZE - DATA_SYMBOLS;
const GENERATOR_BASE = 0;

const EXP = new Uint8Array(FIELD_SIZE * 2);
const LOG = new Int16Array(256);
LOG.fill(-1);
let value = 1;
for (let exponent = 0; exponent < FIELD_SIZE; exponent += 1) {
  EXP[exponent] = value;
  LOG[value] = exponent;
  value <<= 1;
  if (value & 0x100) value ^= PRIMITIVE;
}
for (let exponent = FIELD_SIZE; exponent < EXP.length; exponent += 1) EXP[exponent] = EXP[exponent - FIELD_SIZE];

function multiply(a, b) {
  if (a === 0 || b === 0) return 0;
  return EXP[LOG[a] + LOG[b]];
}

function divide(a, b) {
  if (b === 0) throw new RangeError('GF division by zero');
  if (a === 0) return 0;
  let exponent = LOG[a] - LOG[b];
  if (exponent < 0) exponent += FIELD_SIZE;
  return EXP[exponent];
}

function polynomialMultiply(left, right) {
  const output = new Uint8Array(left.length + right.length - 1);
  for (let i = 0; i < left.length; i += 1) {
    for (let j = 0; j < right.length; j += 1) output[i + j] ^= multiply(left[i], right[j]);
  }
  return output;
}

let generator = Uint8Array.of(1);
for (let index = 0; index < PARITY_SYMBOLS; index += 1) {
  generator = polynomialMultiply(generator, Uint8Array.of(1, EXP[GENERATOR_BASE + index]));
}

function syndromes(codeword) {
  const output = new Uint8Array(PARITY_SYMBOLS);
  for (let index = 0; index < PARITY_SYMBOLS; index += 1) {
    const point = EXP[GENERATOR_BASE + index];
    let result = 0;
    for (const symbol of codeword) result = multiply(result, point) ^ symbol;
    output[index] = result;
  }
  return output;
}

function encodeBlock(data) {
  if (!(data instanceof Uint8Array) || data.length !== DATA_SYMBOLS) throw new RangeError(`RS data block must contain ${DATA_SYMBOLS} symbols`);
  const work = new Uint8Array(FIELD_SIZE);
  work.set(data);
  for (let index = 0; index < DATA_SYMBOLS; index += 1) {
    const coefficient = work[index];
    if (coefficient === 0) continue;
    for (let term = 1; term < generator.length; term += 1) work[index + term] ^= multiply(generator[term], coefficient);
  }
  const output = new Uint8Array(FIELD_SIZE);
  output.set(data);
  output.set(work.subarray(DATA_SYMBOLS), DATA_SYMBOLS);
  return Buffer.from(output);
}

function evaluateAscending(polynomial, point) {
  let result = 0;
  for (let index = polynomial.length - 1; index >= 0; index -= 1) result = multiply(result, point) ^ polynomial[index];
  return result;
}

function berlekampMassey(sequence) {
  // Locator coefficients are stored in ascending power order.
  let locator = Uint8Array.of(1);
  let previous = Uint8Array.of(1);
  let degree = 0;
  let shift = 1;
  let previousDiscrepancy = 1;

  for (let n = 0; n < sequence.length; n += 1) {
    let discrepancy = sequence[n];
    for (let index = 1; index <= degree; index += 1) {
      if (index < locator.length) discrepancy ^= multiply(locator[index], sequence[n - index]);
    }
    if (discrepancy === 0) {
      shift += 1;
      continue;
    }

    const saved = locator.slice();
    const scale = divide(discrepancy, previousDiscrepancy);
    const needed = previous.length + shift;
    if (locator.length < needed) {
      const expanded = new Uint8Array(needed);
      expanded.set(locator);
      locator = expanded;
    }
    for (let index = 0; index < previous.length; index += 1) locator[index + shift] ^= multiply(scale, previous[index]);

    if (2 * degree <= n) {
      degree = n + 1 - degree;
      previous = saved;
      previousDiscrepancy = discrepancy;
      shift = 1;
    } else {
      shift += 1;
    }
  }
  return { locator, degree };
}

function solveMagnitudes(positions, sequence) {
  const count = positions.length;
  const matrix = Array.from({ length: count }, () => new Uint8Array(count + 1));
  for (let row = 0; row < count; row += 1) {
    for (let column = 0; column < count; column += 1) {
      const power = (FIELD_SIZE - 1 - positions[column]) * row;
      matrix[row][column] = row === 0 ? 1 : EXP[power % FIELD_SIZE];
    }
    matrix[row][count] = sequence[row];
  }

  for (let column = 0; column < count; column += 1) {
    let pivot = column;
    while (pivot < count && matrix[pivot][column] === 0) pivot += 1;
    if (pivot === count) throw new Error('RS error magnitude matrix is singular');
    if (pivot !== column) [matrix[pivot], matrix[column]] = [matrix[column], matrix[pivot]];
    const inverse = divide(1, matrix[column][column]);
    for (let item = column; item <= count; item += 1) matrix[column][item] = multiply(matrix[column][item], inverse);
    for (let row = 0; row < count; row += 1) {
      if (row === column || matrix[row][column] === 0) continue;
      const scale = matrix[row][column];
      for (let item = column; item <= count; item += 1) matrix[row][item] ^= multiply(scale, matrix[column][item]);
    }
  }
  return Uint8Array.from(matrix, (row) => row[count]);
}

function decodeBlock(input) {
  if (!(input instanceof Uint8Array) || input.length !== FIELD_SIZE) throw new RangeError(`RS codeword must contain ${FIELD_SIZE} symbols`);
  const output = Uint8Array.from(input);
  let check = syndromes(output);
  if (check.every((symbol) => symbol === 0)) return { data: Buffer.from(output.subarray(0, DATA_SYMBOLS)), correctedSymbols: 0 };

  const { locator, degree } = berlekampMassey(check);
  if (degree < 1 || degree > PARITY_SYMBOLS / 2) throw new Error('RS codeword has too many errors');
  const positions = [];
  for (let position = 0; position < FIELD_SIZE; position += 1) {
    const symbolPower = FIELD_SIZE - 1 - position;
    const inversePoint = EXP[(FIELD_SIZE - symbolPower) % FIELD_SIZE];
    if (evaluateAscending(locator, inversePoint) === 0) positions.push(position);
  }
  if (positions.length !== degree) throw new Error('RS error locations could not be resolved');
  const magnitudes = solveMagnitudes(positions, check);
  for (let index = 0; index < positions.length; index += 1) output[positions[index]] ^= magnitudes[index];
  check = syndromes(output);
  if (!check.every((symbol) => symbol === 0)) throw new Error('RS correction failed');
  return { data: Buffer.from(output.subarray(0, DATA_SYMBOLS)), correctedSymbols: positions.length };
}

function encode(data) {
  if (!Buffer.isBuffer(data)) data = Buffer.from(data);
  const blockCount = Math.ceil(data.length / DATA_SYMBOLS);
  if (blockCount === 0) return { interleaved: Buffer.alloc(0), blockCount: 0 };
  const codewords = [];
  for (let block = 0; block < blockCount; block += 1) {
    const message = Buffer.alloc(DATA_SYMBOLS);
    data.copy(message, 0, block * DATA_SYMBOLS, Math.min(data.length, (block + 1) * DATA_SYMBOLS));
    codewords.push(encodeBlock(message));
  }
  const interleaved = Buffer.alloc(blockCount * FIELD_SIZE);
  let offset = 0;
  for (let column = 0; column < FIELD_SIZE; column += 1) {
    for (let block = 0; block < blockCount; block += 1) interleaved[offset++] = codewords[block][column];
  }
  return { interleaved, blockCount };
}

function decode(interleaved, blockCount, dataLength) {
  if (!Buffer.isBuffer(interleaved)) interleaved = Buffer.from(interleaved);
  if (!Number.isSafeInteger(blockCount) || blockCount < 0 || interleaved.length !== blockCount * FIELD_SIZE) throw new RangeError('Invalid interleaved RS length');
  if (!Number.isSafeInteger(dataLength) || dataLength < 0 || dataLength > blockCount * DATA_SYMBOLS) throw new RangeError('Invalid decoded RS length');
  const decoded = Buffer.alloc(blockCount * DATA_SYMBOLS);
  let correctedSymbols = 0;
  for (let block = 0; block < blockCount; block += 1) {
    const codeword = Buffer.alloc(FIELD_SIZE);
    for (let column = 0; column < FIELD_SIZE; column += 1) codeword[column] = interleaved[column * blockCount + block];
    const result = decodeBlock(codeword);
    result.data.copy(decoded, block * DATA_SYMBOLS);
    correctedSymbols += result.correctedSymbols;
  }
  for (const value of decoded.subarray(dataLength)) if (value !== 0) throw new Error('RS padding is not zero');
  return { data: decoded.subarray(0, dataLength), correctedSymbols };
}

module.exports = {
  PRIMITIVE,
  FIELD_SIZE,
  DATA_SYMBOLS,
  PARITY_SYMBOLS,
  GENERATOR_BASE,
  encodeBlock,
  decodeBlock,
  encode,
  decode,
};
